pkg/lib/cockpit-upload-helper.ts

```diff
@@ -0,0 +1,133 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2024 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import cockpit from 'cockpit';
+
+function debug(...args: unknown[]) {
+    if (window.debugging == 'all' || window.debugging?.includes('upload'))
+        console.debug('debug[upload]', ...args);
+}
+
+class UploadError extends Error {
+    name = 'UploadError';
+    constructor(problem: string, message: cockpit.JsonObject) {
+        // XXX message.message is untranslated.  Should we just always use cockpit.message(problem)?
+        super(typeof message.message === 'string' ? message.message : cockpit.message(problem));
+    }
+}
+
+/* XXX: these values are ~a decade old and were more or less copied from ssh
+ * (where they were last changed in 2007).  We might test alternatives...
+ */
+const BLOCK_SIZE = 16 << 10; // 16kiB
+const FLOW_WINDOW = 2 << 20; // 2MiB
+
+export async function upload(
+    destination: string,
+    stream: ReadableStream,
+    progress?: (bytes_sent: number) => void,
+    signal?: AbortSignal,
+    options?: cockpit.JsonObject
+) {
+    /* While the upload is running, the function is essentially always waiting
+     * inside of a Promise resolved by notify().  Anything that might cause
```

Comment:
This is confusing: Promises are *not* resolved by `.notify()`. That provides progress messages, the final resolving/rejection *ends* the .notify() stream.

Reply:
See below -- please rename "notify" to something that makes it clear that it is a Promise resolve callback.